### PR TITLE
added new rule for defining action

### PR DESCRIPTION
**Pull request by eturan**
According comment lib/virgilio:15 //To prevent accidental overwrites, all internal variables end with a `$`-sign.
also action names should be checked to prevent putting $ sign to end of action name.

It is not possible to add namespace end with '$'. For example,  you cannot add namespace with following code. It will throw 'InvalidNamespaceError'.

virgilio.namespace$('animal$.human').defineAction$(function speak() {
    return 'Hello world!';
});

In order to make it consistent, same rule should be applied for action names. For example you would be able to define action before this pull request

virgilio.namespace$('animal.human').defineAction$(function speak$() {
    return 'Hello world!';
});

But it conflicts with Virgilio's own rule.

So with this pull request, we apply same rule also for action names. If user tries to define action end with $ sign, Virgilio will throw 'InvalidActionNameError'

### CODE REVIEW

File: lib/virgilio.js [hunk 1]
```diff
@@ -141,7 +144,7 @@ Virgilio.prototype.namespace$ = function namespace$(path) {
     //Follow the namespace stack, each element a child of the previous element.
     while ((name = namespaceStack.pop())) {
         namespace = namespace._getChildNamespace$(name) ||
-                                    namespace._createNamespace$(name);
+        namespace._createNamespace$(name);
```
Comment (lucamaraschi) [inline lib/virgilio.js]: **why**?

**Comment by lucamaraschi** [comment]
Why should a $ ending action be forbidden?

**Comment by eturan** [comment]
Why cannot you create namespace ending with $? Same rule should apply for actions.

Also, you could define action called 'defineAction$' and after defined this action, original defineAction$ will be overridden. So original defineAction$ function could be useless in this case.

**Comment by jwoudenberg** [comment]
Actually, the behavior for the namespaces wasn't written with the explicit goal of forbidding $-ending namespaces, bot to ensure people don't pick namespaces like 'foo.....bar' that will cause parsing issues. Not allowing $-signs was a side-effect of that change, though arguably it makes sense to.

I'm conflicted about this supposed change, two questions:

- Virgilio is meant to be easily hackable/extensible. Should extensions be allowed to register their own $-sign methods and, if so, how should they preferably go about it?
- Why use the namespace regex for validating action names?

**Comment by lucamaraschi** [comment]
To followup to what @jwoudenberg is saying, I would like to challenge the solution of implementing it as a custom linter rule in a gulp action: in this way we would just alert people of their (miss)behavior, keeping the principle of Virgilio respected and adding a nice step of explicit type safety...

**Comment by eturan** [comment]
@jwoudenberg 
* Extensions should be able to register their own ‘$’ method, because it’s an extension of a Virgilio method. But looking at the code, ‘$extend’ is actually overriding the method. Not extending. By calling super$ we can extend the method$, but by default it’s not a extension.

* this reggex /^\w+(\.\w+)*$/ makes sures we use a-zA-X0-9 character set. Maybe too restrictive. We can give more freedom and allow other characters like _. Only $ at the end of the word should not be allowed (to avoid overlapping with virgilio’s methods). But I used the namespace regex to reuse code and because that regex was also meant to avoid $ on namespaces, I imagined it would also make sense to apply it on actions

@lucamaraschi 
I don't agree creating custom linter rule in gulp action since it may effect using another libraries with Virgilio. If it is needed to implement type safety to this repository, I would use Flow (http://flowtype.org/)

**Comment by lucamaraschi** [comment]
@eturan Are you really sure?The problem is not type safety in the meaning of checking the type consistency (as JS is a dynamic language) but just to make sure that overriding of Virgilioś methods is alerted as a wanrning to the developer...and I can see it implemented in a very easy way without altering the bahavior of other consuming libraries.

**Comment by eturan** [comment]
@lucamaraschi yes we can alter by using for example a simple plugin (https://www.npmjs.com/package/gulp-check/). 

As you said Flow is a completely different solution and I think it is not necessary as well.

**Pull request closed without merging**